### PR TITLE
add TLS auth for accessing Prometheus

### PR DESCRIPTION
**Pull request by iyalang**
This PR allows to use TLS key pair (together with CA) for authentication when accessing Prometheus that is outside the cluster. The TLS cert is optional and TLS auth is skipped in case only CA file path is provided.

### CODE REVIEW

**Comment by iyalang** [comment]
Could anyone have a look please? :)

File: cmd/adapter/adapter.go [hunk 1]
```diff
@@ -351,6 +359,21 @@ func makePrometheusCAClient(caFilename string) (*http.Client, error) {
 		return nil, fmt.Errorf("no certs found in prometheus-ca-file")
 	}
 
+	if (tlsCertFilename != "") && (tlsKeyFilename != "") {
+		tlsClientCerts, err := tls.LoadX509KeyPair(tlsCertFilename, tlsKeyFilename)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read TLS key pair: %s", err.Error())
```
Comment (dgrisonnet) [inline cmd/adapter/adapter.go:365]: nit: 
```suggestion
			return nil, fmt.Errorf("failed to read TLS key pair: %v", err)
```

**Review by dgrisonnet — COMMENTED** [review]
Hi @iyalang, thank you for your contribution :slightly_smiling_face:, and sorry for the delay.

Could you please sign the CNCF CLA? We are in the process of moving this project to kubernetes-sigs and would want to make sure we met all the necessary prerequisite.

/cc @s-urbaniak can you please have a look at this PR?

File: cmd/adapter/adapter.go (hunk 1 above)
Comment (iyalang) [reply cmd/adapter/adapter.go]: changed

**Comment by iyalang** [comment]
Thank you! I signed the CLA yesterday.

**Review by dgrisonnet — APPROVED** [review]
Thanks!

**Comment by s-urbaniak** [comment]
lgtm as well, just wondering if we can have some sort of testing around this?

**Comment by s-urbaniak** [comment]
And thank you @iyalang for the contribution! :raised_hands: :tada:

**Comment by iyalang** [comment]
Sure, I can write tests, but I probably won't be able to do it within the next days.

**Comment by s-urbaniak** [comment]
No worries at all :)

**Comment by iyalang** [comment]
@s-urbaniak I added tests for the HTTP server and solved the CLA issue. Sorry for the delay.

File: cmd/adapter/adapter.go [hunk 2]
```diff
@@ -120,6 +124,10 @@ func (cmd *PrometheusAdapter) addFlags() {
 		"kubeconfig file used to configure auth when connecting to Prometheus.")
 	cmd.Flags().StringVar(&cmd.PrometheusCAFile, "prometheus-ca-file", cmd.PrometheusCAFile,
 		"Optional CA file to use when connecting with Prometheus")
+	cmd.Flags().StringVar(&cmd.PrometheusTLSCertFile, "prometheus-tls-cert-file", cmd.PrometheusTLSCertFile,
+		"Optional TLS cert file to use when connecting with Prometheus")
+	cmd.Flags().StringVar(&cmd.PrometheusTLSKeyFile, "prometheus-tls-key-file", cmd.PrometheusTLSKeyFile,
+		"Optional TLS key file to use when connecting with Prometheus")
```
Comment (s-urbaniak) [inline cmd/adapter/adapter.go:130]: nit: let's document that key/certificate rotation is not handled here and must be implemented elsewhere, i.e. a sidecar.

Comment (s-urbaniak) [reply cmd/adapter/adapter.go:130]: another nit: as this refers to client certificates, can we call this `prometheus-client-tls-cert-file` and `prometheus-client-tls-key-file`?

**Comment by s-urbaniak** [comment]
@iyalang awesome, thank you! I just have two nits which I would love to have addressed. Thank you!

File: cmd/adapter/adapter.go (hunk 2 above)
Comment (iyalang) [reply cmd/adapter/adapter.go]: Changed

**Comment by iyalang** [comment]
@s-urbaniak implemented the changes

**Review by dgrisonnet — CHANGES_REQUESTED** [review]
Great work with the tests! I just have one comment, could you perhaps increase the validity of the certificates? Currently they are only valid until 2022, which means that in a year the tests will be failing.

**Comment by iyalang** [comment]
@dgrisonnet That's a very good point! I updated all valid certs, now they should expire in ~2700 years :)

**Comment by dgrisonnet** [comment]
Thanks you for the great work on this PR! :slightly_smiling_face: :rocket: 

/lgtm

**Comment by s-urbaniak** [comment]
Thank you @iyalang ! :tada:

**Pull request merged**